### ISSUE TITLE
[processor/resourcedetection] Proposal: default/optional resource attributes per detector

### ISSUE THREAD

**Issue by dmitryax** [open-telemetry/opentelemetry-collector-contrib#21482]
### Is your feature request related to a problem? Please describe.

Currently, it's possible to configure the list of added resource attributes by the processor, for example the following config reduces the list of added attributes:

```yaml
resourcedetection:
  detectors: [ec2, system]
  attributes: [host.id, host.name, cloud.provider]
```

However, since multiple detectors set the same resource attributes with different values, users cannot set one attribute from one detector and another attribute from another detector using a single resourcedetection processor instance. In such cases, the first detector in the list always wins and sets all the attributes.

Specifically, according to [the spec definition](https://github.com/open-telemetry/opentelemetry-specification/blob/88e0e7e5b408aabc855180156cebbe202a7fa081/specification/resource/semantic_conventions/host.md?plain=1#LL13C1-L13C1), `host.name` should preferably be obtained from the host, while `host.id` should preferably be obtained from the cloud provider. Therefore, it is not possible to use a single resource detection processor to obtain `host.name` and `host.id` from their respective preferred sources.

### Describe the solution you'd like

Given that most of the detectors already have their own configurations sections, the suggestion is to make the resource attributes configurable with the same configuration interface as scraping receivers have, e.g. [apache receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/apachereceiver/internal/metadata/generated_metrics.go#L94-L102)

This allows us to have optional and default attributes. The suggestion is to define default/optional attributes accordingly to the spec recommendation and avoid any overlap between default attributes used in the host and cloud detectors. 

To address the issue with obtaining `host.id` and `host.name` from their preferred sources, the default configuration for the `ec2` and `system` detectors should be as follows:
```
resourcedetection:
  system:
    resource_attributes:
      host.name:
        enabled: true
      host.id:
        enabled: false
  ec2:
    resource_attributes:
      host.name:
        enabled: false
      host.id:
        enabled: true
```

Users still will be able to change the source for the attribute values.

**Comment by dloucasfx** [comment]
@dmitryax 
this a great feature! it gives the user more flexibility to pick and choose their attributes.
If this was accepted, is there a reason to keep https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/config.go#L43

**Comment by dmitryax** [comment]
> If this was accepted, is there a reason to keep https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/config.go#L43

No, I believe this should deprecate the Attributes config option

**Comment by omrozowicz-splunk** [comment]
I can work on this, please assign it to me.

**Comment by mx-psi** [comment]
@omrozowicz-splunk do you have any updates on this? I am interested in making this happen so I am happy to help/take over/review :)

**Comment by omrozowicz-splunk** [comment]
Hey @mx-psi, actually I've just started working on this and I'm targeting to have a draft end of this week/beginning of the next week so I can inform you here when it's done. Appreciate any help/input 😃

**Comment by mx-psi** [comment]
@omrozowicz-splunk awesome :) feel free to ping me on the PR to review. My only comment is that I wonder if we can leverage `metadata.yaml` to define a list of resource attributes and reuse the work for #16373 here (but this does not need to be the first version)

**Comment by omrozowicz-splunk** [comment]
@mx-psi yeah, so I try to reuse the `resource_attributes` logic, this is the only case I've seen that we have "nested" `metadata.yaml`, so I put it this way in `doc.go` of `resourcedetectionprocessor`:
```
//go:generate mdatagen metadata.yaml
//go:generate mdatagen internal/aws/ec2/metadata.yaml
```
and it indeed created the correct structure, default config and so on, but now I'm struggling a bit with configuring it from `processors` perspective.
For example, when I configure:
```
processors:
  resourcedetection:
    detectors: [ec2, system]
    ec2:
      resource_attributes:
        host.name:
          enabled: true
```
then the structs sets all the fields to false except `host.name` and on the Detector level it is impossible to differentiate if `resource_attribute` was set to be disabled or just wasn’t provided. Maybe I miss some understanding of it, or the detector's `resource_attributes` config shouldn't be:
```
type Config struct {
...
	ResourceAttributes metadata.ResourceAttributesConfig `mapstructure:"resource_attributes"`
}
```
But changing it to something else makes `ResourceAttributesConfig` useless in this case 🤷‍♀️

**Comment by mx-psi** [comment]
> on the Detector level it is impossible to differentiate if resource_attribute was set to be disabled or just wasn’t provided

Not sure if I understand this, what would be the difference in behavior between 'unset' and disabled? If something should be enabled by default when unset, the way to go about this is to set the field to `true` on the `createDefaultConfig` function here https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3771c6532550410930e5a02c5fcef4f7795f6ba0/processor/resourcedetectionprocessor/factory.go#L84

I am going to be out next week, but I think @dmitryax can help with this if this doesn't help or you have any further questions

**Comment by mx-psi** [comment]
I am back, happy to further help if you have other questions or want feedback on anything @omrozowicz-splunk :)

**Comment by omrozowicz-splunk** [comment]
Hey @mx-psi, just opened a draft. Sorry so late, but I was off a few days this week as well :) We already had some initial review on my repo (https://github.com/omrozowicz-splunk/opentelemetry-collector-contrib/pull/3) if you want to take a look. 
Still a few things requires work:
1. So far every attribute is set to `true`, will need to assess what should be enabled by default
2. Adjust unit tests to code changes
3. I think I messed up something with those `x.CreateDefaultConfig()` in `resourcedetectionprocessor/config.go`, because it's still not being updated by the overwritten values. Comparing to my previous problem now at least it gets the default structure by default.

Will continue work on this next week, but please put a comment on Draft if you're in the mood to review all these files and want to help 😄

**Comment by mx-psi** [comment]
This was fixed by #23253
